src/Control/Eff/Fresh.hs

```diff
@@ -25,18 +33,40 @@ import Control.Eff
 -- | Create unique Enumerable values.
 data Fresh v where
   Fresh :: Fresh Int
+  Replace :: !Int -> Fresh ()
+
+instance ( MonadBase m m
+         , Typeable m
+         , SetMember Lift (Lift m) r
+         , MonadBaseControl m (Eff r)
+         ) => MonadBaseControl m (Eff (Fresh ': r)) where
+    type StM (Eff (Fresh ': r)) a = StM (Eff r) (a, Int)
+    liftBaseWith f = do i <- fresh
+                        raise $ liftBaseWith $ \runInBase ->
+                          f (\k -> runInBase $ runFreshReturn k i)
+    restoreM x = do (r,i) <- raise (restoreM x)
+                    replace i
+                    pure r
```

Comment:
@greydot there's a build error on `7.8.4` due to this file (might also be in other cases. something like the below should fix it
```haskell
#if __GLASGOW_HASKELL__ < 710
import Control.Applicative
#endif
```